### PR TITLE
Create a read-optimized implementation of IModel

### PR DESCRIPTION
**Pull request by AndriySvyryd**
Part of #8258

### CODE REVIEW

File: src/EFCore/Metadata/SlimNavigation.cs [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+using Microsoft.EntityFrameworkCore.Metadata.Internal;
+
+#nullable enable
+
+namespace Microsoft.EntityFrameworkCore.Metadata
+{
+    /// <summary>
+    ///     Represents a navigation property which can be used to navigate a relationship.
+    /// </summary>
+    public class SlimNavigation : SlimPropertyBase, INavigation
```
Comment (smitpatel) [inline src/EFCore/Metadata/SlimNavigation.cs:19]: So we don't have runtimenavigation, runtimeproperty, runtimeindex? Are we planning to add it?

File: src/EFCore/Metadata/SlimEntityType.cs [hunk 2]
```diff
@@ -0,0 +1,1366 @@
...
+        private SlimKey? FindPrimaryKey()
+            => _baseType?.FindPrimaryKey() ?? _primaryKey;
+
+        /// <summary>
+        ///     Sets the primary key for this entity type.
+        /// </summary>
+        /// <param name="key"> The new primary key. </param>
+        public virtual void SetPrimaryKey([NotNull] SlimKey key)
+        {
+            foreach (var property in key.Properties)
```
Comment (smitpatel) [inline src/EFCore/Metadata/SlimEntityType.cs:152]: no checks for basetype being null?

File: src/EFCore/Metadata/SlimModel.cs [hunk 3]
```diff
@@ -0,0 +1,259 @@
...
+    ///     <para>
+    ///         This is a light-weight implementation that is constructed from a built model and is not meant to be used at design-time.
+    ///     </para>
+    /// </summary>
+    public class SlimModel : AnnotatableBase, IRuntimeModel
+    {
+        private readonly SortedDictionary<string, SlimEntityType> _entityTypes = new(StringComparer.Ordinal);
+        private readonly ConcurrentDictionary<Type, PropertyInfo?> _indexerPropertyInfoMap = new();
+        private readonly ConcurrentDictionary<Type, string> _clrTypeNameMap = new();
+        private readonly Dictionary<Type, SortedSet<SlimEntityType>> _sharedTypes = new() { };
```
Comment (smitpatel) [inline src/EFCore/Metadata/SlimModel.cs:36]: is "{ }" required?

**Review by smitpatel — APPROVED** [review]
:shipit:

File: src/EFCore/Properties/CoreStrings.resx [hunk 4]
```diff
@@ -1411,6 +1412,9 @@
   <data name="SkipNavigationWrongType" xml:space="preserve">
     <value>The skip navigation '{navigation}' cannot be removed from the entity type '{entityType}' because it is defined on the entity type '{otherEntityType}'.</value>
   </data>
+  <data name="SlimModelMissingData" xml:space="preserve">
+    <value>The requested configuration is not stored in the read-optimized model, please use 'DbContext.DesignTimeModel'.</value>
```
Comment (roji) [inline src/EFCore/Properties/CoreStrings.resx:1416]:
```suggestion
    <value>The requested configuration is not stored in the slim model, please use 'DbContext.DesignTimeModel'.</value>
```

File: src/EFCore/Metadata/SlimNavigation.cs (hunk 1 above)
Comment (AndriySvyryd) [reply src/EFCore/Metadata/SlimNavigation.cs:19]: No, we plan to remove them. Currently tracked as a Backlog task in https://github.com/dotnet/efcore/issues/1906

File: src/EFCore/Metadata/SlimEntityType.cs (hunk 2 above)
Comment (AndriySvyryd) [reply src/EFCore/Metadata/SlimEntityType.cs:152]: These types are meant to be constructed from validated models only, therefore any checks would be redundant

File: src/EFCore/Properties/CoreStrings.resx (hunk 4 above)
Comment (AndriySvyryd) [reply src/EFCore/Properties/CoreStrings.resx:1416]: I think the original has more meaning

Comment (roji) [reply src/EFCore/Properties/CoreStrings.resx:1416]: OK, but shouldn't we be consistent between our messages and the actual type naming etc.?

Comment (AndriySvyryd) [reply src/EFCore/Properties/CoreStrings.resx:1416]: Sure, but I don't want to name the type `ReadOptimizedModel`

**Pull request merged**